docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -397,6 +397,8 @@ The following components are supported:
 - **SyntheticGlobal[**\ `name`\ **]** selects the synthetic global with name `name`.
 - **ArrayElement** selects the elements of an array.
 - **Element** selects the elements of a collection-like container.
+- **WithoutElement** selects a collection-like container without its elements (input only).
+- **WithElement** selects the elements of a collection-like container but not the container itself (input only).
```

Comment:
Maybe I should re-phrase this. For Java, `A.Element` and `A.WithElement` are almost the same, except that `A.WithElement` will (access path wise) point to `A`, but flow wise it will only include flow from the elements of `A` (corresponding to `A.Element`). That is `A.WithElement` still selects `A`, but omits flow from `A` itself and only includes flow from the `Elements` of `A`.
However, according to @hvitved  there might be a performance benefit in declaring summaries as `A.WithElement -> B` instead of `A.Element -> B.Element`.

Reply:
Lets say you are making a shallow copy (or clone) of a collection then, a summary could look something like `Argument[this].WithElement -> ReturnValue`

Reply:
So, is this criteria correct?
* When extracting individual elements from a collection, `.Element` should be used. Thas is, when the output node has the type of the elements of the collection.
* When extracting all elements from a collection (like when doing a copy), `.WithElement` should be used. That is, when the output node has a collection-like type.

Reply:
That sounds like a good rule of thumb. @hvitved: Do you have any other recommendation?

Reply:
Yes, that's a good rule of thumb.